Add Mosquitto tuning (#1747)
theforeman/foreman-documentation: guides/common/modules/proc_rex-pull-based-provider.adoc

[id="REX_pull_based_provider_{context}"]

> = REX pull based provider 

Reviewer: ```suggestion
= Remote Execution pull-based transport
```

> Remote Execution has been around for a while now and has served SSH-based execution very well.

Reviewer: This is to define the acronym before we use it to make it clear what we are talking about.

 
Now, {Project} has a ‘pull’ based provider that has the features of REX but with a deployment model similar to katello-agent. 

This provider uses MQTT as its messaging protocol and includes an MQTT client running on each content host.

> In order to tune the MQTT server (`mosquitto`) and increase the number of clients (content hosts) connnected to it, you can follow the following procedure on the {Project} server:

Reviewer: ```suggestion
In order to tune the MQTT server (`mosquitto`) and increase the number of clients connected to it, you can follow the following procedure on the {Project} server:
```

.Procedure

> . Enable the feature using: `{foreman-installer} --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt` (this will change the REX transport mode from the default SSH one to MQTT).
>   It's important to mention that one {Project} server ({ProjectServer} or {SmartProxyServer}) can only use one transport mode: SSH or MQTT.

Reviewer 2: ```suggestion
. Enable pull-based remote execution on your {ProjectServer} or {SmartProxyServer}:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
{foreman-installer} --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
----
+
Note that your {ProjectServer} or {SmartProxyServer} can only use one transport mode, either SSH or MQTT.
```

Maybe reword slightly here and there; but in general, we use code blocks instead of inline commands.

> 

Reviewer 2: ```suggestion
```

> . To increase the default number of clients accepted by the MQTT server, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` service to handle ~5000 clients):

Reviewer: ```suggestion
. To increase the default number of hosts accepted by the MQTT server, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` service to handle ~5000 clients):
```

Reviewer 2: Is this optional? What's the default value? Is 5000 clients a recommendation? Does it depends on the specs/used tuning profile?

Reviewer 3: It is optional, but there might be dragons, depending on your scale. According to systemd docs[1] the default limit is 1024. Based on the reasoning in[1], I wouldn't say it depends on specs/tuning profile.

[1] - https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Process%20Properties

Reviewer 4: Please see https://github.com/theforeman/foreman-documentation/pull/1747#discussion_r1009725326 for more discussion. I'd say that if we raise beyond 5000, we might as well raise it WAY higher. Then if we do want to limit it, we should use `max_connections` in `mosquitto.conf` instead.

Reviewer 5: We chose 5000 because it's what we were able to test, but agree with your reasoning.

+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
cat >/etc/systemd/system/mosquitto.service.d/limits.conf <<__EOF__
[Service]
LimitNOFILE=5000
__EOF__
----

> . Once the file has been edited, the following commands need to be run to make the tuning come into effect:

Reviewer 2: ```suggestion
. Run the following commands apply your changes:
```

Reviewer 5: s/apply/to apply/

Author: done. thank you.

+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----

> systemctl daemon-reload
> systemctl restart mosquitto.service

Reviewer 2: ```suggestion
# systemctl daemon-reload
# systemctl restart mosquitto.service
```
